cloudsync_onedrive.py

```diff
@@ -1197,43 +1161,61 @@ def hash_data(self, file_like) -> str:
                 h.update(c)
             return h.hexdigest().upper()
 
-    @property                                # type: ignore
-    def namespace(self) -> str:              # type: ignore
-        return self._drive_id_to_name(self._drive_id) if self._drive_id else None
+    @property
+    def namespace(self) -> Optional[Namespace]:
+        return self._namespace
 
     @namespace.setter
-    def namespace(self, ns: str):
-        ns_id = self._drive_name_to_id(ns)
-        if not ns_id:
-            raise CloudNamespaceError("Unknown namespace %s" % ns)
-
-        if ns_id != self._drive_id:
-            log.debug("namespace changing to %s", ns)
-
-        self.namespace_id = ns_id
+    def namespace(self, ns: Namespace):
+        self.namespace_id = ns.id
 
     @property
     def _is_biz(self):
         if self.__cached_is_biz is None:
-            dat = self._direct_api("get", "/drives/%s/" % self._drive_id)
+            dat = self._direct_api("get", "/drives/%s/" % self._validated_namespace_id)
             self.__cached_is_biz = dat["driveType"] != 'personal'
         return self.__cached_is_biz
 
+    @property
+    def _validated_namespace_id(self):
+        if self.connected and self.namespace_id:
+            return self.namespace_id
+        raise CloudNamespaceError("namespace_id has not been validated")
+
     @property
     def namespace_id(self) -> Optional[str]:
-        return self._drive_id
+        return self._namespace.id if self._namespace else None
 
     @namespace_id.setter
     def namespace_id(self, ns_id: str):
-        self._drive_id = ns_id
+        if self.connected:
+            # validate
+            self._fetch_drive_list()
+            drive = self.__drive_by_id.get(ns_id, None)
+            if not drive:
+                try:
+                    api_drive = self._direct_api("get", f"/drives/{ns_id}/")
+                    # Note: name is generic here (e.g., "Documents") because we don't know the parent site
+                    drive = self._save_drive_info(api_drive["name"], ns_id)
+                except CloudFileNotFoundError:
+                    self._namespace = None
+                    log.error("Failed to find namespace: %s", ns_id)
+                    raise CloudNamespaceError("Failed to find namespace")
+            self._namespace = drive
+            log.info("namespace validated")
+        else:
+            # defer validation until a connection is established
+            self._namespace = self.__drive_by_id.get(ns_id, Namespace(name=ns_id, id=ns_id))
+            log.info("namespace validation deferred")
+        log.info("USING NS name=%s id=%s", self.namespace.name, self.namespace_id)
```

Comment:
How frequently does this occur? Are we sure we want to info log here?

Reply:
namespace setting should only happen once per connection -- note that the USING NS log statement was moved from connect_impl().

I'll roll the other 2 log.info calls into this statement to keep everything together.